File: rd-kt/rd-framework/src/main/kotlin/com/jetbrains/rd/framework/impl/AsyncRdSet.kt
package com.jetbrains.rd.framework.impl

import com.jetbrains.rd.framework.*
import com.jetbrains.rd.framework.base.*
import com.jetbrains.rd.util.lifetime.Lifetime
import com.jetbrains.rd.util.reactive.IScheduler
import com.jetbrains.rd.util.reactive.IViewableSet
import com.jetbrains.rd.util.string.RName
import com.jetbrains.rd.util.threading.SynchronousScheduler
import java.util.function.*

class AsyncRdSet<T : Any> private constructor(
    private val set: BackendRdSet<T>
) : IRdBindable,  MutableSet<T>, IAsyncSource2<IViewableSet.Event<T>>{

    constructor(valueSerializer: ISerializer<T> = Polymorphic()) : this(BackendRdSet(valueSerializer))

    companion object : ISerializer<AsyncRdSet<*>> {
        fun <T : Any> read(ctx: SerializationCtx, buffer: AbstractBuffer, valueSerializer: ISerializer<T> = Polymorphic()): AsyncRdSet<T> = AsyncRdSet(valueSerializer).withId(RdId.read(buffer))
        override fun write(ctx: SerializationCtx, buffer: AbstractBuffer, value: AsyncRdSet<*>) = value.rdid.write(buffer)

        override fun read(ctx: SerializationCtx, buffer: AbstractBuffer): AsyncRdSet<*> {
            return read(ctx, buffer, Polymorphic())
        }
    }

    val change: IAsyncSource2<IViewableSet.Event<T>> = object : IAsyncSource2<IViewableSet.Event<T>> {
        override fun adviseOn(lifetime: Lifetime, scheduler: IScheduler, handler: (IViewableSet.Event<T>) -> Unit) {
            set.change.advise(lifetime) {
                scheduler.queue { handler(it) }
            }
        }
    }

    override fun adviseOn(lifetime: Lifetime, scheduler: IScheduler, handler: (IViewableSet.Event<T>) -> Unit) {
        synchronized(set) {
            set.advise(lifetime) { event ->
                scheduler.queue { handler(event) }
            }
        }
    }

    fun withId(id: RdId) : AsyncRdSet<T> {
        synchronized(set) {
            set.withId(id)
        }
        return this
    }

    var optimizeNested: Boolean
        get() = true
        set(value) {}

    var async: Boolean
        get() = true
        set(value) {}

    override val rdid: RdId
        get() = synchronized(set) {
            set.rdid
        }

    override fun preBind(lf: Lifetime, parent: IRdDynamic, name: String) {
        synchronized(set) {
            set.preBind(lf, parent, name)
        }
    }

    override fun bind() {
        synchronized(set) {
            set.bind()
        }
    }

    override fun identify(identities: IIdentities, id: RdId) {
        synchronized(set) {
            set.identify(identities, id)
        }
    }

    override fun deepClone(): IRdBindable {
        synchronized(set) {
            val clone = set.deepClone()
            return AsyncRdSet(clone)
        }
    }

    override val protocol: IProtocol?
        get() = set.protocol

    override val serializationContext: SerializationCtx?
        get() = set.serializationContext

    override val location: RName
        get() = set.location

    override val size: Int
        get() = set.size

    override fun add(element: T): Boolean {
        return synchronized(set) {
            set.add(element)
        }
    }

    override fun addAll(elements: Collection<T>): Boolean {
        return synchronized(set) {
            set.addAll(elements)
        }
    }

    override fun clear() {
        synchronized(set) {
            set.clear()
        }
    }

    override fun iterator(): MutableIterator<T> {
       val iterator =  set.iterator()

        return object : MutableIterator<T> {
            override fun hasNext(): Boolean = iterator.hasNext()
            override fun next(): T = iterator.next()

            override fun remove() {
                synchronized(set) {
                    iterator.remove()
                }
            }
        }
    }

    override fun remove(element: T): Boolean {
        return synchronized(set) {
            set.remove(element)
        }
    }

    override fun removeAll(elements: Collection<T>): Boolean {
        return synchronized(set) {
            set.removeAll(elements)
        }
    }

    override fun removeIf(filter: Predicate<in T>): Boolean {
        TODO()
    }

    override fun retainAll(elements: Collection<T>): Boolean {
        return synchronized(set) {
            set.retainAll(elements)
        }
    }

    override fun contains(element: T): Boolean {
        return set.contains(element)
    }

    override fun containsAll(elements: Collection<T>): Boolean {
        return set.containsAll(elements)
    }

    override fun isEmpty(): Boolean {
        return set.isEmpty()
    }

    private class BackendRdSet<T : Any>(
        valueSerializer: ISerializer<T>
    ) : RdSet<T>(valueSerializer) {

        init {
            async = true
        }

        override fun assertBindingThread() {
        }

        override fun assertThreading() {
        }

        override fun onWireReceived(proto: IProtocol, buffer: AbstractBuffer, ctx: SerializationCtx, dispatchHelper: IRdWireableDispatchHelper) {
            synchronized(this) {
                super.onWireReceived(proto, buffer, ctx, object : IRdWireableDispatchHelper {
                    override val rdId: RdId
                        get() = dispatchHelper.rdId
                    override val lifetime: Lifetime
                        get() = dispatchHelper.lifetime

                    override fun dispatch(lifetime: Lifetime, scheduler: IScheduler?, action: () -> Unit) {
                        dispatchHelper.dispatch(lifetime, SynchronousScheduler, action)
                    }
                })
            }
        }

        override fun deepClone(): BackendRdSet<T> = BackendRdSet(valueSerializer).also { for (elem in it) { it.add(elem.deepClonePolymorphic()) } }
    }
}

File: rd-net/RdFramework/Impl/AsyncRdSet.cs
using System;
using System.Collections;
using System.Collections.Generic;
using JetBrains.Collections.Viewable;
using JetBrains.Diagnostics;
using JetBrains.Lifetimes;
using JetBrains.Rd.Base;
using JetBrains.Rd.Util;
using JetBrains.Serialization;

namespace JetBrains.Rd.Impl;

public class AsyncRdSet<T> : IRdBindable, IAsyncSource<SetEvent<T>>, 
#if NET35
    ICollection<T>
#else
  ISet<T>, IReadOnlyCollection<T>
#endif 
  where T : notnull
{
  private readonly RdSetBackend mySet;
  private readonly AsyncSignal<SetEvent<T>> mySignal = new();

  public AsyncRdSet(CtxReadDelegate<T> readValue, CtxWriteDelegate<T> writeValue)
  {
    mySet = new RdSetBackend(readValue, writeValue);
    mySet.Change.Advise(Lifetime.Eternal, x => mySignal.Fire(x));
  }

  public IAsyncSource<SetEvent<T>> Change => mySignal;
  
  public bool IsMaster = false;
  
  public RdId RdId
  {
    get
    {
      lock (mySet)
        return mySet.RdId;
    }
    set
    {
      lock (mySet)
        mySet.RdId = value;
    }
  }

  public void PreBind(Lifetime lf, IRdDynamic parent, string name)
  {
    lock (mySet) 
      mySet.PreBind(lf, parent, name);
  }

  public void Bind()
  {
    lock (mySet) 
      mySet.Bind();
  }

  public void Identify(IIdentities identities, RdId id)
  {
    lock (mySet) 
      mySet.Identify(identities, id);
  }

  public bool OptimizeNested
  {
    get => true;
    set { }
  }
  
  public bool ValueCanBeNull
  {
    get => mySet.ValueCanBeNull;
    set => mySet.ValueCanBeNull = value;
  }

  public bool Async
  {
    get => true;
    set { }
  }

  public RName Location
  {
    get
    {
      lock (mySet)
        return mySet.Location;
    }
  }

  public IProtocol? TryGetProto()
  {
    return mySet.TryGetProto();
  }

  public bool TryGetSerializationContext(out SerializationCtx ctx)
  {
    return mySet.TryGetSerializationContext(out ctx);
  }

  public int Count => mySet.Count;
  public bool IsReadOnly => mySet.IsReadOnly;

#if !NET35
  bool ISet<T>.Add(T item)
  {
    lock (mySet)
      return mySet.Add(item);
  }

  public void UnionWith(IEnumerable<T> other)
  {
    lock (mySet) 
      mySet.UnionWith(other);
  }

  public void IntersectWith(IEnumerable<T> other)
  {
    lock (mySet) 
      mySet.IntersectWith(other);
  }

  public void ExceptWith(IEnumerable<T> other)
  {
    lock (mySet) 
      mySet.ExceptWith(other);
  }

  public void SymmetricExceptWith(IEnumerable<T> other)
  {
    lock (mySet) 
      mySet.SymmetricExceptWith(other);
  }

  public bool IsSubsetOf(IEnumerable<T> other)
  {
    return mySet.IsSubsetOf(other);
  }

  public bool IsSupersetOf(IEnumerable<T> other)
  {
    return mySet.IsSubsetOf(other);
  }

  public bool IsProperSupersetOf(IEnumerable<T> other)
  {
    return mySet.IsProperSupersetOf(other);
  }

  public bool IsProperSubsetOf(IEnumerable<T> other)
  {
    return mySet.IsProperSubsetOf(other);
  }

  public bool Overlaps(IEnumerable<T> other)
  {
    return mySet.Overlaps(other);
  }

  public bool SetEquals(IEnumerable<T> other)
  {
    return mySet.Overlaps(other);
  }
#endif

  public void Add(T item)
  {
    lock (mySet)
      mySet.Add(item);
  }

  public void Clear()
  {
    lock (mySet)
      mySet.Clear();
  }

  public bool Contains(T item)
  {
    return mySet.Contains(item);
  }

  public void CopyTo(T[] array, int arrayIndex)
  {
    mySet.CopyTo(array, arrayIndex);
  }

  public bool Remove(T item)
  {
    lock (mySet) 
      return mySet.Remove(item);
  }

  public IEnumerator<T> GetEnumerator() => mySet.GetEnumerator();
  IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();

  public static AsyncRdSet<T> Read(SerializationCtx ctx, UnsafeReader reader)
  {
    return Read(ctx, reader, Polymorphic<T>.Read, Polymorphic<T>.Write);
  }

  public static AsyncRdSet<T> Read(SerializationCtx ctx, UnsafeReader reader, CtxReadDelegate<T> readValue, CtxWriteDelegate<T> writeValue)
  {
    var id = reader.ReadRdId();
    return new AsyncRdSet<T>(readValue, writeValue).WithId(id);
  }

  public static void Write(SerializationCtx ctx, UnsafeWriter writer, AsyncRdSet<T> value)
  {
    Assertion.Assert(!value.RdId.IsNil);
    writer.Write(value.RdId);
  }

  public CtxReadDelegate<T> ReadValueDelegate => mySet.ReadValueDelegate;
  public CtxWriteDelegate<T> WriteValueDelegate => mySet.WriteValueDelegate;

  private class RdSetBackend : RdSet<T>
  {
    public RdSetBackend(CtxReadDelegate<T> readValue, CtxWriteDelegate<T> writeValue) : base(readValue, writeValue)
    {
      OptimizeNested = true;
      Async = true;
    }
    
    protected override void AssertBindingThread()
    {
    }

    protected override void AssertThreading()
    {
      
    }

    public override void OnWireReceived(IProtocol proto, SerializationCtx ctx, UnsafeReader stream, IRdWireableDispatchHelper dispatchHelper)
    {
      lock (this)
        base.OnWireReceived(proto, ctx, stream, new DelegatingDispatchHelper(dispatchHelper));
    }
  }
  
  private class DelegatingDispatchHelper : IRdWireableDispatchHelper
  {
    private readonly IRdWireableDispatchHelper myDispatchHelper;
    
    public RdId RdId => myDispatchHelper.RdId;
    public Lifetime Lifetime => myDispatchHelper.Lifetime;

    public DelegatingDispatchHelper(IRdWireableDispatchHelper dispatchHelper)
    {
      myDispatchHelper = dispatchHelper;
    }
    
    public void Dispatch(IScheduler? scheduler, Action action)
    {
      myDispatchHelper.Dispatch(SynchronousScheduler.Instance, action);
    }
  }

  public void Print(PrettyPrinter printer)
  {
    lock (mySet) 
      mySet.Print(printer);
  }

  public void AdviseOn(Lifetime lifetime, IScheduler scheduler, Action<SetEvent<T>> action)
  {
    lock (mySet)
    {
      mySet.Advise(lifetime, e =>
      {
        scheduler.Queue(() => action(e));
      });
    }
  }
}